test/Grpc.AspNetCore.Server.Tests/PipeExtensionsTests.cs

```diff
@@ -0,0 +1,355 @@
+        {
+            // Arrange
+            var ms = new MemoryStream(new byte[]
+                {
+                    0x00, // compression = 0
+                    0x00,
+                    0x00
+                });
+
+            var pipeReader = new StreamPipeReader(ms);
+
+            // Act
+            var ex = Assert.ThrowsAsync<InvalidOperationException>(
+                () => pipeReader.ReadMessageAsync().AsTask());
+
+            // Assert
+            Assert.AreEqual("Incomplete message.", ex.Message);
+        }
+
+        [Test]
+        public void ReadMessageAsync_MessageDataIncomplete_ThrowError()
+        {
+            // Arrange
+            var ms = new MemoryStream(new byte[]
+                {
+                    0x00, // compression = 0
+                    0x00,
+                    0x00,
+                    0x00,
+                    0x02, // length = 2
+                    0x10
+                });
+
+            var pipeReader = new StreamPipeReader(ms);
+
+            // Act
+            var ex = Assert.ThrowsAsync<InvalidOperationException>(
+                () => pipeReader.ReadMessageAsync().AsTask());
+
+            // Assert
+            Assert.AreEqual("Incomplete message.", ex.Message);
+        }
+
+        [Test]
+        public void ReadMessageStreamAsync_HeaderIncomplete_ThrowError()
+        {
+            // Arrange
+            var ms = new MemoryStream(new byte[]
+                {
+                    0x00, // compression = 0
+                    0x00,
+                    0x00
+                });
+
+            var pipeReader = new StreamPipeReader(ms);
+
+            // Act
+            var ex = Assert.ThrowsAsync<InvalidOperationException>(
+                () => pipeReader.ReadMessageAsync().AsTask());
+
+            // Assert
+            Assert.AreEqual("Incomplete message.", ex.Message);
+        }
+
+        [Test]
+        public void ReadMessageStreamAsync_MessageDataIncomplete_ThrowError()
+        {
+            // Arrange
+            var ms = new MemoryStream(new byte[]
+                {
+                    0x00, // compression = 0
+                    0x00,
+                    0x00,
+                    0x00,
+                    0x02, // length = 2
+                    0x10
+                });
+
+            var pipeReader = new StreamPipeReader(ms);
+
+            // Act
+            var ex = Assert.ThrowsAsync<InvalidOperationException>(
+                () => pipeReader.ReadMessageStreamAsync().AsTask());
+
+            // Assert
+            Assert.AreEqual("Incomplete message.", ex.Message);
+        }
+
+        [Test]
+        public async Task WriteMessageAsync_NoFlush_WriteNoData()
+        {
+            // Arrange
+            var ms = new MemoryStream();
+            var pipeWriter = new StreamPipeWriter(ms);
+
+            // Act
+            await pipeWriter.WriteMessageAsync(Array.Empty<byte>());
```

Comment:
To actually test not flushing writes no data, the argument should be a non-empty array.

Reply:
An empty message would still write the header.